File: 이하영/240526/10799. 쇠막대기/쇠막대기.py
# 1. ( 나오면 스택에 push
# 2. ) 나오면 (공통적으론 스택 pop )  ( ()는 레이저거나 막대의 처음과 끝을 알리는 것이므로 제외해야 하니까 )
# 2-1. 직전 괄호가 ) 라면 레이저가 나타났다는 뜻이므로 (수만큼 잘림, 스택에 (가 push 된 만큼 카운팅(스택의 길이)
# 2-2. )라면 마지막이라는 의미. +1 카운팅

laser = input()
stack = []
count = 0

for i in range (len(laser)):
    if laser[i] == "(":
        stack.append(laser[i])
    elif laser[i] == ")":
        stack.pop()
        if laser[i-1] == "(":
            count += len(stack)
        else:
            count += 1

print(count)

File: 한민선/240526/10799. 쇠막대기/쇠막대기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        // 1. 레이저를 쏘면 스택의 사이즈 만큼 더해준다 = 레이저를 쏘고 나눠진 조각 수
        // 2. 막대가 끝나면 +1 = 레이저를 쏜 다음 남은 1조각
        Stack<Character> stack = new Stack<>();
        int answer = 0;

        char[] input = br.readLine().toCharArray();
        for (int i = 0; i < input.length; i++) {
            if(input[i] == '('){
                stack.push(input[i]);
            }else{
                stack.pop();
                if(input[i-1] == '('){
                    answer += stack.size();
                }else{
                    answer += 1;
                }
            }
        }
        bw.write(answer + "");
        bw.flush();
        bw.close();
    }
}